### PR TITLE
Pbs nodefile setting (HPC-6960)

### PR DESCRIPTION
**Pull request by itkovian**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,13 @@
-all: env-test.so
+all: env-test.so pbs_nodefile.so
 
 env-test.so: env-test.c
 	gcc -std=gnu99 -Wall -o env-test.o -fPIC -c env-test.c
 	gcc -shared -o env-test.so env-test.o
 
+pbs_nodefile.so: pbs_nodefile.c
```
Comment (stdweird) [inline Makefile]: stop copy/paste? at the very least use some variables for compiling etc

File: pbs_nodefile.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+SPANK_PLUGIN(env-test, 1);
+
+// Gobals
+static char *generate_pbsnodefile_command;
+int init_opts = 0;
+int init_complete = 0;
+
+
+int slurm_spank_init(spank_t sp, int ac, char **av)
```
Comment (stdweird) [inline pbs_nodefile.c]: there is no cleanup of the file when the job completed/failed/....?

Comment (itkovian) [reply pbs_nodefile.c]: Of course there is. It's in the local tmp. Everything there is destroyed.

Comment (stdweird) [reply pbs_nodefile.c]: ? only if you use the private tmpdir plugin. if you can, make this plugin standalone and cleanup the file

**Comment by itkovian** [comment]
~~~~
vsc40075@test2802 (banette) ~/vsc-testing/slurm> squeue
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          67118965   banette slurm_te vsc40075  R       2:24      2 node[2801-2802]
~~~~

node2801:
~~~~
[root@node2801 ~]# ls /tmp/slurm.67118965.0/tmp/
AQXFqFCL1g  ZItz2vyhP8
[root@node2801 ~]# cat /tmp/slurm.67118965.0/tmp/AQXFqFCL1g
node2801
node2801
node2801
node2801
node2802
node2802
node2802
node2802
~~~~

node2802:
~~~~
[root@node2802 ~]# cat /tmp/slurm.67118965.0/tmp/6Lj5lBMmFl
node2801
node2801
node2801
node2801
node2802
node2802
node2802
node2802
~~~~

Output of the job script env:
~~~~
vsc40075@test2802 (banette) ~/vsc-testing/slurm> grep PBS_NOD slurm-67118965.out
PBS_NODENUM=0
PBS_NODEFILE=/tmp/AQXFqFCL1g
~~~~

**Comment by itkovian** [comment]
Still need to figure out how to avoid the duplicate entry, but this at least works and should be usable.

**Comment by itkovian** [comment]
The file is created twice, not sure how to address that. In any case, it should be of no concern, as the job environment will point to a/the correct file.

~~~~
[2018-04-18T09:52:56.394] [67118968.batch] debug:  spank: pbs_nodefile: PBS_NODELIST available in /tmp/w7zfX1f5CD
<snip>
[2018-04-18T09:52:56.578] [67118968.0] debug:  spank: pbs_nodefile: PBS_NODELIST available in /tmp/EOIOwjJWVl
~~~~

File: bdist_rpm.sh [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
-VERSION=$(grep "Version:.*[0-9]" slurm-env-test.spec | tr -s " " |  awk '{print $2;}')
-RELEASE=$(grep "%global rel.*[-1-9]" slurm-env-test.spec | tr -s " " | awk '{print $3}')
+VERSION=$(grep "Version:.*[0-9]" slurm-spank-talamini.spec | tr -s " " |  awk '{print $2;}')
```
Comment (stdweird) [inline bdist_rpm.sh]: use a variable for the project name, so we can reuse the script more often?

File: pbs_nodefile.c [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * (C) 2018 Ghent University
+ *
+ * Author: Andy Georges
+ *
+ * Loosely based on Spank plugin private-tmpdir (c) HPC2N.umu.se
+ *
+ * This plugin generates the PBS_NODEFILE in the $TMPDIR of the job. We assume
+ * other plugins or job epilog will clean these up, so we do not remove files
+ * ourselves.
```
Comment (stdweird) [inline pbs_nodefile.c:34]: add a TODO or know issues wrt the double file

**Comment by stdweird** [comment]
i'm not sure how you managed to do this, but you hav a merge conflict

**Comment by stdweird** [comment]
@itkovian still conflict on bdist_rpm

File: pbs_nodefile.c (hunk 4 above)
Comment (itkovian) [reply pbs_nodefile.c:34]: Done.

File: pbs_nodefile.c [hunk 5]
```diff
@@ -0,0 +1,182 @@
+/*
+ * Plugin to set the PBS_NODEFILE env variable and its contents for the job
+ *
+ * (C) 2018 Ghent University
```
Comment (wdpypere) [inline pbs_nodefile.c]: this is not our standard license header block.

File: Makefile [hunk 6]
```diff
@@ -1,8 +1,20 @@
...
+
+%.o: %.c
+	$(CC) $(CFLAGS) -o $@ -c $<
+
+%.so: %.o
+	$(CC) $(LDFLAGS) $< -o $@
 
 clean:
-	rm -f env-test.o env-test.so
+	rm -f $(OBJECTS) $(SHLIBS)
```
Comment (wdpypere) [inline Makefile]: missing newline

File: TODO [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Known issues
+
+- Both the <job>.batch and the job.0, job.1, ... processes create the PBS_NODEFILE and change the jobs' environment.
```
Comment (stdweird) [inline TODO:3]: already open an issue with schedmd so they can have a look how to do this properly (ie what method to call) once we have a contract.

**Comment by itkovian** [comment]
Issue with SchedMD: https://bugs.schedmd.com/show_bug.cgi?id=5273

File: pbs_nodefile.c (hunk 5 above)
Comment (itkovian) [reply pbs_nodefile.c]: Of course not. Code is based on other code, so I need to attribute.

File: TODO (hunk 7 above)
Comment (itkovian) [reply TODO:3]: Done.

File: Makefile (hunk 6 above)
Comment (itkovian) [reply Makefile]: Done.

**Comment by wdpypere** [comment]
@itkovian this good to go?

**Comment by itkovian** [comment]
@wdpypere This is already deployed, so yes :)

**Pull request merged**